### PR TITLE
Add state transform for compute.ForwardingRule

### LINKED ISSUE #2471: gcp.compute.ForwardingRule breaks on latest pulumi/gcp 7.38.0   & pulumi/gcp  8.3.1
### Describe what happened

When we update the Pulumi GCP provider from version 6 to either version 7 (7.38.0) or version 8 (8.3.1), we encounter an unusual error with the ForwardingRule resource during the pulumi preview step:

``` gcp:compute:ForwardingRule (internal-lb-fw-rule):
    error:   sdk-v2/provider2.go:509: sdk.helper_schema: missing expected {: provider=google-beta@8.3.1
    error: diffing urn:pulumi:dev::web-3-loadbalancer::avrotros:InternalLoadBalancer$gcp:compute/forwardingRule:ForwardingRule::internal-lb-fw-rule: 1 error occurred:
        * [service_directory_registrations[0]] missing expected {
```

Everything works correctly when using Pulumi GCP version 6.67.1."

### Sample program

problematic resource inside a ComponentResource:

```      
        this.forwardingRule = new gcp.compute.ForwardingRule(
            `${name}-fw-rule`,
            {
                target: this.regionHttpProxy.id,
                portRange: "80",
                region: args.region,
                loadBalancingScheme: "INTERNAL_MANAGED",
                network: args.networkId,
                subnetwork: this.subNetwork.id,
                allowGlobalAccess: true,
            },
            { parent: this },
        );
```
### Log output

when using pulumi/gcp 8.3.1

``` gcp:compute:ForwardingRule (internal-lb-fw-rule):
    error:   sdk-v2/provider2.go:509: sdk.helper_schema: missing expected {: provider=google-beta@8.3.1
    error: diffing urn:pulumi:dev::web-3-loadbalancer::avrotros:InternalLoadBalancer$gcp:compute/forwardingRule:ForwardingRule::internal-lb-fw-rule: 1 error occurred:
        * [service_directory_registrations[0]] missing expected {
```

when using pulumi/gcp 7.38.0

```gcp:compute:ForwardingRule (internal-lb-fw-rule):
    error:   sdk-v2/provider2.go:385: sdk.helper_schema: missing expected {: provider=google-beta@7.38.0
    error: diffing urn:pulumi:dev::web-3-loadbalancer::avrotros:InternalLoadBalancer$gcp:compute/forwardingRule:ForwardingRule::internal-lb-fw-rule: 1 error occurred:
        * [service_directory_registrations[0]] missing expected {
```

### Affected Resource(s)

_No response_

### Output of `pulumi about`

when using pulumi/gcp  8.3.1

```CLI          
Version      3.134.1
Go Version   go1.23.1
Go Compiler  gc

Plugins
KIND      NAME    VERSION
language  nodejs  unknown

Host     
OS       ubuntu
Version  22.04
Arch     x86_64

This project is written in nodejs: executable='/home/clenn/.nvm/versions/node/v20.11.1/bin/node' version='v20.11.1'
```

when using  pulumi/gcp  7.38.0

```CLI          
Version      3.134.1
Go Version   go1.23.1
Go Compiler  gc

Plugins
KIND      NAME    VERSION
resource  gcp     7.38.0
language  nodejs  unknown

Host     
OS       ubuntu
Version  22.04
Arch     x86_64

This project is written in nodejs: executable='/home/clenn/.nvm/versions/node/v20.11.1/bin/node' version='v20.11.1'`
```

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by guineveresaenger**
This pull request addressses a gap in state migration from v6 of this provider to modern versions.
The affected field is serviceDirectoryRegistrations, which changed from a List to an Object.
This change will allow for a successful upgrade from v6.67.1 to the next release of this provider.

Fixes #2471.

- **Add state migration to endure compute.ForwardingRule can upgrade from v6 to mordern versions**
- **Add upgrade test for compute forwarding rule**

### CODE REVIEW

File: provider/resources.go [hunk 1]
```diff
@@ -989,12 +989,31 @@ func Provider() tfbridge.ProviderInfo {
...
+				Tok: gcpResource(gcpCompute, "ForwardingRule"),
+				TransformFromState: func(_ context.Context, pMap resource.PropertyMap) (resource.PropertyMap, error) {
+					// The service_directory_registrations field changes from [] in v6 to an object in v7.
+					// It appears that on early versions of v7, upstream has state migrators; they are lost by v7.38
+					// of this provider. See https://github.com/pulumi/pulumi-gcp/issues/2471.
+					pValue := resource.NewProperty(pMap)
+					serviceDirRegistrationsPath := resource.PropertyPath{"serviceDirectoryRegistrations"}
+					val, ok := serviceDirRegistrationsPath.Get(pValue)
+					if ok && val.IsArray() {
+						if len(val.ArrayValue()) > 0 {
```
Comment (VenelinMartinov) [inline provider/resources.go:1007]: I don't believe this branch is covered by tests, can we please add a test for a non-nil serviceDirectoryRegistrations?
You can use https://github.com/pulumi/providertest/pull/95 like in https://github.com/pulumi/pulumi-azure/blob/12949f95dacdda96ffd9d7710d73bc1b48bda565/provider/provider_yaml_test.go#L79

File: provider/resources.go [hunk 2]
```diff
@@ -989,12 +989,31 @@ func Provider() tfbridge.ProviderInfo {
...
+			"google_compute_disk":                        {Tok: gcpResource(gcpCompute, "Disk")},
+			"google_compute_firewall":                    {Tok: gcpResource(gcpCompute, "Firewall")},
+			"google_compute_firewall_policy":             {Tok: gcpResource(gcpCompute, "FirewallPolicy")},
+			"google_compute_firewall_policy_rule":        {Tok: gcpResource(gcpCompute, "FirewallPolicyRule")},
+			"google_compute_firewall_policy_association": {Tok: gcpResource(gcpCompute, "FirewallPolicyAssociation")},
+			"google_compute_forwarding_rule": {
+				Tok: gcpResource(gcpCompute, "ForwardingRule"),
+				TransformFromState: func(_ context.Context, pMap resource.PropertyMap) (resource.PropertyMap, error) {
+					// The service_directory_registrations field changes from [] in v6 to an object in v7.
+					// It appears that on early versions of v7, upstream has state migrators; they are lost by v7.38
```
Comment (VenelinMartinov) [inline provider/resources.go:1001]: Can you please link these state migrators? I have not been able to find any.

Is this not a MaxItemsOne change? If so, this should have been caught by the type checker in the bridge, as the type of the property has changed. This might be an effect of PRC, I'll try to test this in https://github.com/pulumi/pulumi-terraform-bridge/issues/2418

**Review by VenelinMartinov — COMMENTED** [review]
Thanks for fixing, can you please add the missing test coverage for the state migrator?

This might also be a bridge issue with the type checker.

File: provider/resources.go (hunk 2 above)
Comment (VenelinMartinov) [reply provider/resources.go:1001]: Actually, this probably never worked for the type checker - it only checks the inputs, not the state.

Comment (VenelinMartinov) [reply provider/resources.go:1001]: This is likely https://github.com/pulumi/pulumi-terraform-bridge/issues/1667 rather - potentially made worse by PRC as we now save and expect the empty list in state for block types.

Comment (guineveresaenger) [reply provider/resources.go:1001]: The only evidence I have here is what I said in the original bug issue - the upgrade works v6 -> v7 -> v8 but not v6 -> v8. It's entirely possible that our rollout of continued support for maxItemsOne is responsible somehow - v6 is two major versions back at this point.

File: provider/provider_yaml_test.go [hunk 3]
```diff
@@ -142,6 +142,21 @@ func TestConnectionProfileUpgradev7(t *testing.T) {
...
 
+func TestComputeForwardingRuleV6V8StateMigration(t *testing.T) {
+	result := testUpgrade(t, "test-programs/compute-forwardingrule")
+	assertpreview.HasNoReplacements(t, result)
+	assertpreview.HasNoDeletes(t, result)
+}
+
+func TestComputeForwardingRuleV6V8StateMigrationNoDefaults(t *testing.T) {
+	t.Parallel()
+	if testing.Short() {
```
Comment (VenelinMartinov) [inline provider/provider_yaml_test.go:153]: testUpgrade takes care of this it is not necessary here.

**Review by VenelinMartinov — COMMENTED** [review]
Tests are still failing, I've restarted in case it's a flake, PTAL.

**Comment by guineveresaenger** [comment]
Tests are failing because tests don't clean up after they fail. I've requested access to clean up resources manually.

**Comment by VenelinMartinov** [comment]
The cluster test 409 should be fixed in https://github.com/pulumi/pulumi-gcp/pull/2516, try merging that to see if it resolves the clash?

**Review by iwahbe — APPROVED** [review]

**Review by VenelinMartinov — APPROVED** [review]
